backend/auth/clients.ts

```diff
@@ -0,0 +1,162 @@
+import { ApiContext } from "."
+import { requireAuth } from "../util/validateAuth"
+import { Client } from "@prisma/client"
+
+const crypto = require("crypto")
+
+export function createClient(ctx: ApiContext) {
+  return async (req: any, res: any) => {
+    let auth = await requireAuth(req.headers.authorization, ctx)
+    if (auth.error) {
+      return res.status(403).json({ error: auth })
+    }
+    if (!auth.admin) {
+      return res.status(403).json({ error: "You do not have permission." })
+    }
+
+    let name = req.body.name
+    let redirect_uri = req.body.redirect_uri
+
+    let client_id = crypto.randomBytes(6).toString("hex")
+    let client_secret = crypto.randomBytes(64).toString("hex")
+
+    let client = (
+      await ctx
+        .knex("clients")
+        .insert({
+          name,
+          client_id,
+          client_secret,
+          redirect_uri,
+        })
+        .returning("*")
+    )?.[0]
+
+    return res.status(200).json({
+      success: true,
+      client,
+    })
+  }
+}
+
+export function getClients(ctx: ApiContext) {
+  return async (req: any, res: any) => {
+    let auth = await requireAuth(req.headers.authorization, ctx)
+    if (auth.error) {
+      return res.status(403).json({ error: auth })
+    }
+    if (!auth.admin) {
+      return res.status(403).json({ error: "You do not have permission." })
+    }
+
+    let clients = await ctx.knex.select<any, Client[]>("*").from("clients")
+
+    return res.status(200).json(clients)
+  }
+}
+
+export function showClient(ctx: ApiContext) {
+  return async (req: any, res: any) => {
+    let auth = await requireAuth(req.headers.authorization, ctx)
+    if (auth.error) {
+      return res.status(403).json({ error: auth })
+    }
+    if (!auth.admin) {
+      return res.status(403).json({ error: "You do not have permission." })
+    }
+
+    const id = req.query.id
```

Comment:
Could we use `req.params.id` in here instead of query, so that the request would be `/api/client/1` instead of `/api/client?id=1`? Now you can do `/api/client` without any parameters -- in this case, it would return 404 but as this pattern seems to repeat in all GET requests here. But better be safe -- if we were to refactor this to use `prisma`, for example, it could just query _all_ the clients and return the first one.